sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala

```diff
@@ -79,20 +80,30 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
       case s if s.isEmpty =>
         SetCommand(None)
       case _ => throw new ParseException("Expected format is 'SET', 'SET key', or " +
-        "'SET key=value'. If you want to include special characters in key, " +
-        "please use quotes, e.g., SET `ke y`=value.", ctx)
+        "'SET key=value'. If you want to include special characters in key or semicolon is value," +
+        " please use quotes, e.g., SET `ke y`=`value`.", ctx)
     }
   }
 
-  override def visitSetQuotedConfiguration(ctx: SetQuotedConfigurationContext)
-    : LogicalPlan = withOrigin(ctx) {
-    val keyStr = ctx.configKey().getText
-    if (ctx.EQ() != null) {
-      SetCommand(Some(keyStr -> Option(remainder(ctx.EQ().getSymbol).trim)))
-    } else {
-      SetCommand(Some(keyStr -> None))
+  override def visitSetQuotedConfiguration(
+      ctx: SetQuotedConfigurationContext): LogicalPlan = withOrigin(ctx) {
+      if (ctx.configValue() != null && ctx.configKey() != null) {
+        SetCommand(Some(ctx.configKey().getText -> Option(ctx.configValue().getText)))
+      } else if (ctx.configValue() != null) {
+        SetCommand(Some(ctx.anything.getText -> Option(ctx.configValue().getText)))
```

Comment:
Shall we check the key here to make sure it matches `configKeyDef`?

Reply:
oh right~